Content/Items/Misc/Weapons.Soilgun.cs

```diff
@@ -0,0 +1,1488 @@
+            Vector2 shootVelocity = Utils.SafeNormalize(Projectile.velocity, Vector2.UnitY) * shootSpeed;
+            for (int i = 0; i < 4 + Main.rand.Next(3); i++)
+            {
+                Projectile.NewProjectile(Projectile.GetSource_FromThis(), position, (shootVelocity.RotatedByRandom(MathHelper.ToRadians(18))) * Main.rand.NextFloat(0.9f, 1.1f), ModContent.ProjectileType<SoilGunSoilProjs>(), damage, knockBack, owner.whoAmI, SoilType);
+            }
+            for (int i = 0; i < 15; i++)
+            {
+                Vector2 dustVelocity = shootVelocity.RotatedByRandom(MathHelper.ToRadians(8)) * Main.rand.NextFloat(0.25f, 0.45f);
+                Dust dust = Dust.NewDustDirect(position, 2, 8, DustID.Dirt, dustVelocity.X, dustVelocity.Y);
+                dust.scale = Main.rand.NextFloat(1.1f, 1.55f);
+                dust.noGravity = true;
+            }
+            for (int i = 0; i < 7; i++)
+            {
+                Vector2 sandVelocity = (shootVelocity * Main.rand.NextFloat(0.12f, 0.20f)).RotatedByRandom(MathHelper.ToRadians(9f));
+                Dust.NewDustDirect(position, 2, 8, ModContent.DustType<Dusts.Sand>(), sandVelocity.X, sandVelocity.Y, 140, default, 0.5f);
+
+                Dust.NewDustDirect(position, 2, 8, ModContent.DustType<SandNoGravity>(), sandVelocity.X, sandVelocity.Y, 135, default, 0.65f);
+
+                Dust.NewDustDirect(position, 2, 8, ModContent.DustType<SandNoGravity>(), sandVelocity.X, sandVelocity.Y, 135, default, 0.55f);
+            }
+            CameraSystem.Shake += 2;
+            owner.reuseDelay = 15;
+            Item item = owner.GetModPlayer<SoilgunPlayer>().SoilItem;
+            item.stack--;
+            if (item.stack <= 1)
+            {
+                item.TurnToAir();
+            }
+            //need a better sfx
+            SoundEngine.PlaySound(SoundID.Item61, Projectile.position);
+        }
+    }
+
+    class SandNoGravity : ModDust
+    {
+        public override string Texture => AssetDirectory.Dust + "Sand";
+
+        public override void OnSpawn(Dust dust)
+        {
+            dust.noGravity = true;
+            dust.noLight = true;
+            dust.scale *= 6;
+            dust.frame = new Rectangle(0, 0, 10, 10);
+        }
+
+        public override Color? GetAlpha(Dust dust, Color lightColor)
+        {
+            return dust.color;
+        }
+
+        public override bool Update(Dust dust)
+        {
+            dust.color = Lighting.GetColor((int)(dust.position.X / 16), (int)(dust.position.Y / 16)).MultiplyRGB(Color.White) * 0.2f * (dust.alpha / 255f);
+            dust.position += dust.velocity;
+            dust.scale *= 0.9745f;
+            dust.velocity *= 0.97f;
+            dust.rotation += 0.1f;
+
+            if (dust.scale <= 0.2f)
+                dust.active = false;
+            return false;
+        }
+    }
+
+    class SoilGunSoilProjs : ModProjectile, IDrawPrimitive
```

Comment:
This should very likely be an abstract class with unique classes for each soil type as they all have different behavior. Having one godclass like this is poor practice as it makes both readability and scalability worse.